### ISSUE TITLE
Remote participant doesn't receive audio when local participant changes input device

### ISSUE THREAD

**Issue by TimWillaert** [twilio/twilio-video.js#1479]
- [x] I have verified that the issue occurs with the latest twilio-video.js release and is not marked as a known issue in the [CHANGELOG.md](https://github.com/twilio/twilio-video.js/blob/master/CHANGELOG.md).
 - [x] I reviewed the [Common Issues](https://github.com/twilio/twilio-video.js/blob/master/COMMON_ISSUES.md) and open GitHub issues and verified that this report represents a potentially new issue.
 - [x] I verified that the [Quickstart](https://github.com/twilio/video-quickstart-js) application works in my environment.
 - [x] I am not sharing any [Personally Identifiable Information (PII)](https://www.twilio.com/docs/glossary/what-is-personally-identifiable-information-pii)
       or sensitive account information (API keys, credentials, etc.) when reporting this issue.

**Code to reproduce the issue:**

```js

// Function to change input device

const setDevice = (device) => {
        if(!room) return
        let deviceId = device.deviceId
        const localParticipant = room.localParticipant
        if(device.kind === 'audioinput'){
            setSelectedAudioDevice(device.label)
            Video.createLocalAudioTrack({
                deviceId: {exact: deviceId}
            }).then((localAudioTrack) => {
                const tracks = Array.from(localParticipant.audioTracks.values()).map(
                    function(trackPublication) {
                        return trackPublication.track;
                    });
                localParticipant.unpublishTracks(tracks);
                localParticipant.publishTrack(localAudioTrack);
            })
        } else if(device.kind === 'videoinput'){
            setSelectedVideoDevice(device.label)
            Video.createLocalVideoTrack({
                deviceId: {exact: deviceId}
            }).then((localVideoTrack) => {
                const tracks = Array.from(localParticipant.videoTracks.values()).map(
                    function(trackPublication) {
                        return trackPublication.track;
                    });
                localParticipant.unpublishTracks(tracks);
                localParticipant.publishTrack(localVideoTrack);
            })
        }
    }

// Part of Participant.js component

const trackpubsToTracks = (trackMap) =>
    Array.from(trackMap.values())
      .map((publication) => publication.track)
      .filter((track) => track !== null);

  useEffect(() => {
    setVideoTracks(trackpubsToTracks(participant.videoTracks));
    setAudioTracks(trackpubsToTracks(participant.audioTracks));

    const trackSubscribed = (track) => {
      if (track.kind === "video") {
        setVideoTracks((videoTracks) => [...videoTracks, track]);
      } else if (track.kind === "audio") {
        setAudioTracks((audioTracks) => [...audioTracks, track]);
      }
    };

    const trackUnsubscribed = (track) => {
      if (track.kind === "video") {
        setVideoTracks((videoTracks) => videoTracks.filter((v) => v !== track));
      } else if (track.kind === "audio") {
        setAudioTracks((audioTracks) => audioTracks.filter((a) => a !== track));
      }
    };

    // This trackPublished function is to update the local participant's (the person who changed input devices) own view, as the local participant and remote participants share the same component

    const trackPublished = (track) => {
      if(track.kind === 'video' && track.track?.isFromSelf === true){
        setVideoTracks([track.track]);
      } else if(track.kind === 'audio' && track.track?.isFromSelf === true) {
        setAudioTracks([track.track]);
      }
    }

    participant.on("trackSubscribed", trackSubscribed);
    participant.on("trackUnsubscribed", trackUnsubscribed);
    participant.on("trackPublished", trackPublished)

    return () => {
      setVideoTracks([]);
      setAudioTracks([]);
      participant.removeAllListeners();
    };
  }, [participant]);
```

**Expected behavior:**

Other people should still hear a user when that user changes their input device.

**Actual behavior:**

This code works perfectly fine for changing the video input device. A user can change their video device and all remote participants will see the new camera. However, it doesn't work for audio devices. When a user changes their audio device, all other participants can no longer hear that user.

**Software versions:**

 - [x] Browser(s): Chrome
 - [x] Operating System: Windows 10
 - [x] twilio-video.js: 2.14.0
 - [x] Third-party libraries (e.g., Angular, React, etc.): React

Edit: I started from this article: https://www.twilio.com/blog/video-chat-react-hooks

**Comment by charliesantos** [comment]
Hey @TimWillaert , thanks for submitting! The article you referenced seems to be old. Can you please check out our [reference react app](https://github.com/twilio/twilio-video-app-react) instead? 
I believe this should address the issue you mentioned. You can start looking in this [section](https://github.com/twilio/twilio-video-app-react/blob/master/src/components/DeviceSelectionDialog/AudioInputList/AudioInputList.tsx#L20) of the code. The API you need is [track.restart()](https://sdk.twilio.com/js/video/releases/2.14.0/docs/LocalAudioTrack.html#restart__anchor)

**Comment by TimWillaert** [comment]
Hi @charliesantos , thanks for the fast response! The restart method works correctly. However, I still have a small problem. I have an audio visualizer on each Participant so that I can highlight when they are speaking. It seems however that the .restart() is only emitted to remote participants. When the local participant changes their audio device, the audio visualizer works on the remote participant's end, but not for the local participant himself. Is there a way to notify the local participant of this restart as well?

**Comment by charliesantos** [comment]
Hi @TimWillaert , I'm glad .restart() fixed your issue. Regarding the audio visualizer use case you mentioned, I'm not sure if I understand your use case correctly, so let me clarify a few things. For example, you have three participants in a room - Alice, Bob, and Charlie. You want to highlight Alice when she speaks. So while Alice is speaking, who will trigger .restart() and who do you want to get notified when this happens? Is your use case is to just determine who is currently speaking and get notified if the dominant speaker changes?

**Comment by TimWillaert** [comment]
Hi, @charliesantos, my bad for not explaining my use case well enough. If you're familiar with Discord, whenever you're in a group call, anyone who is speaking gets a green border around their icon, and this is visible for all participants. So I'm not highlighting the dominant speaker, I'm visually changing everyone's webcam to have a green border whenever they are speaking. To do this, each participant has an AudioContext and analyser. This is done client-side, so participants are not emitting when they are speaking. Each client is analyzing each participant.

The problem I'm having is the following: It's working fine when everyone joins the room, but let's say Bob changes their audio input device. In that case, every other participant will still correctly see a green border around Bob's webcam when Bob is speaking, however Bob will not see a green border around himself anymore. I'm guessing that the local participant is not notified of the track restart, or something like that?

I have the following Participant.jsx component:
`
const Participant = ({ participant, selfMuted, selfHidden, leader, mutedUsers, hiddenUsers }) => {
  const [videoTracks, setVideoTracks] = useState([]);
  const [audioTracks, setAudioTracks] = useState([]);

  const [audioContext, setAudioContext] = useState(new AudioContext());

  //Speaking indicator
  const [isSpeaking, setIsSpeaking] = useState(false);

  const videoRef = useRef();
  const audioRef = useRef();

  const trackpubsToTracks = (trackMap) =>
    Array.from(trackMap.values())
      .map((publication) => publication.track)
      .filter((track) => track !== null);

  useEffect(() => {
    setVideoTracks(trackpubsToTracks(participant.videoTracks));
    setAudioTracks(trackpubsToTracks(participant.audioTracks));

    const trackSubscribed = (track) => {
      if (track.kind === "video") {
        setVideoTracks((videoTracks) => [...videoTracks, track]);
      } else if (track.kind === "audio") {
        setAudioTracks((audioTracks) => [...audioTracks, track]);
      }
    };

    const trackUnsubscribed = (track) => {
      if (track.kind === "video") {
        setVideoTracks((videoTracks) => videoTracks.filter((v) => v !== track));
      } else if (track.kind === "audio") {
        setAudioTracks((audioTracks) => audioTracks.filter((a) => a !== track));
      }
    };

    participant.on("trackSubscribed", trackSubscribed);
    participant.on("trackUnsubscribed", trackUnsubscribed);

    return () => {
      setVideoTracks([]);
      setAudioTracks([]);
      participant.removeAllListeners();
    };
  }, [participant]);

  useEffect(() => {
    const videoTrack = videoTracks[0];
    if (videoTrack) {
      videoTrack.attach(videoRef.current);
      return () => {
        videoTrack.detach();
      };
    }
  }, [videoTracks]);

  useEffect(() => {
    const audioTrack = audioTracks[0];
    if (audioTrack) {
      audioTrack.attach(audioRef.current);

      let mediaStream = new MediaStream();
      mediaStream.addTrack(audioTrack.mediaStreamTrack);

      let analyser = audioContext.createAnalyser();
      let microphone = audioContext.createMediaStreamSource(mediaStream);
    
      let javascriptNode = audioContext.createScriptProcessor(2048, 1, 1);
    
      analyser.smoothingTimeConstant = 0.3;
      analyser.fftSize = 1024;
    
      microphone.connect(analyser);
      analyser.connect(javascriptNode);
      javascriptNode.connect(audioContext.destination);
    
      let timeout = setTimeout(() => {
        setIsSpeaking(false)
      }, 800)
    
      javascriptNode.onaudioprocess = () => {
        var array =  new Uint8Array(analyser.frequencyBinCount);
        analyser.getByteFrequencyData(array);
        var values = 0;
    
        var length = array.length;
        for (var i = 0; i < length; i++) {
            values += array[i];
        }

        var average = values / length;
        if(average > 20){
            setIsSpeaking(true)
            window.clearTimeout(timeout)
            timeout = setTimeout(() => {
              setIsSpeaking(false)
            }, 800)
        }
      }
      return () => {
        audioTrack.detach();
      };
    }
  }, [audioTracks]);
};
`

**Comment by charliesantos** [comment]
@TimWillaert , it sounds like your use case is already supported by the SDK and that you don't need to implement your own analyser on the client side. I suggest to look at [room.dominantSpeaker](https://sdk.twilio.com/js/video/releases/2.14.0/docs/Room.html) and listen for the event [dominantSpeakerChanged](https://sdk.twilio.com/js/video/releases/2.14.0/docs/Room.html#event:dominantSpeakerChanged__anchor). Our [reference react app](https://github.com/twilio/twilio-video-app-react) is doing something similar but instead of highlighting with green, we move the participants to the top. Please have a look at it. Here's the [react hook](https://github.com/twilio/twilio-video-app-react/blob/master/src/hooks/useDominantSpeaker/useDominantSpeaker.tsx) that implements this.

**Comment by TimWillaert** [comment]
@charliesantos Thanks, but can't there only be one dominant speaker at a time? In my use case, if both Bob and Alice are speaking, both of them should be highlighted.

**Comment by charliesantos** [comment]
@TimWillaert in that case, analyzing audio levels on the remote track is the correct solution. Our reference react app also have this feature where we put audio level indicators instead of a green border. Can you please check it out? I don't think it has the issue you mentioned above this thread.

**Comment by TimWillaert** [comment]
@charliesantos Okay I will, thanks for the help!

**Comment by charliesantos** [comment]
I'm closing this now. Please feel free to follow up if you have further questions @TimWillaert
